qiskit/quantum_info/operators/dihedral/dihedral.py

```diff
@@ -0,0 +1,495 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019, 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+CNOTDihedral operator class.
+
+Methods for working with the CNOT-dihedral group.
+
+Example:
+
+  from dihedral import CNOTDihedral
+  g = CNOTDihedral(3)  # create identity element on 3 qubits
+  g.cnot(0,1)          # apply CNOT from qubit 0 to qubit 1
+  g.flip(2)            # apply X on qubit 2
+  g.phase(3, 1)        # apply T^3 on qubit 1
+  print(g)             # pretty print g
+
+  phase polynomial =
+   0 + 3*x_0 + 3*x_1 + 2*x_0*x_1
+  affine function =
+   (x_0,x_0 + x_1,x_2 + 1)
+
+ This means that |x_0 x_1 x_2> transforms to omega^{p(x)}|f(x)>,
+ where omega = exp(i*pi/4) from which we can read that
+ T^3 on qubit 1 AFTER CNOT_{0,1} is the same as
+ T^3 on qubit 0, T^3 on qubit 1, and CS_{0,1} BEFORE CNOT_{0,1}.
+"""
+
+import itertools
+import numpy as np
+
+from qiskit.exceptions import QiskitError
+from qiskit.quantum_info.operators.base_operator import BaseOperator
+from qiskit.quantum_info.operators.operator import Operator
+from qiskit.quantum_info.operators.pauli import Pauli
+from qiskit.quantum_info.operators.scalar_op import ScalarOp
+from qiskit.quantum_info.synthesis.cnotdihedral_decompose import decompose_cnotdihedral
+from qiskit.circuit import QuantumCircuit, Instruction
+from .dihedral_circuits import _append_circuit
+from .polynomial import SpecialPolynomial
+
+
+class CNOTDihedral(BaseOperator):
+    """CNOT-dihedral Object Class.
+    The CNOT-dihedral group on num_qubits qubits is generated by the gates
+    CNOT, T and X.
+
+    References:
+        1. Shelly Garion and Andrew W. Cross, *Synthesis of CNOT-Dihedral circuits
+           with optimal number of two qubit gates*, `Quantum 4(369), 2020
+           <https://quantum-journal.org/papers/q-2020-12-07-369/>`_
+        2. Andrew W. Cross, Easwar Magesan, Lev S. Bishop, John A. Smolin and Jay M. Gambetta,
+           *Scalable randomised benchmarking of non-Clifford gates*,
+           npj Quantum Inf 2, 16012 (2016).
+    """
+
+    def __init__(self, data, validate=True):
+        """Initialize a CNOTDihedral operator object."""
+
+        # Initialize from another CNOTDihedral by sharing the underlying
+        # poly, linear and shift
+        if isinstance(data, CNOTDihedral):
```

Comment:
Should you be able to initialize directly from a tuple `(linear, shift, poly)` ?

Reply:
I think it will be too complicated, since these are 3 different types (matrix, vector, SpecialPolynomial).